### ISSUE TITLE
Warn if same variable is read and written in same expression

### ISSUE THREAD

**Issue by safinaskar** [rust-lang/rust-clippy#10880]
### What it does

**TL;DR**. Please, warn in code like `a.f(g(&mut a))`.

**Details**. Just now I got very unfortunate situation with my production code. Let me describe it. I parse some custom protobuf-based data format. Using my own code (i. e. without preexisting protobuf libs).

I have this variable (holding file contents):
```rust
let mut uncompressed: &[u8] = ...
```

Then I read one protobuf varint from it so:
```rust
let split_here = read_usize(&mut uncompressed).unwrap();
```

Now `split_here` is integer value I have read. And now `uncompressed` is remaining data.

Now I read `split_here`-sized data so:

```rust
let (archive_info, uncompressed2) = uncompressed.split_at(split_here);
```

Okay. So far, so good. The code above works correctly.

Now I refactored it so:
```rust
let (archive_info, uncompressed2) = uncompressed.split_at(read_usize(&mut uncompressed).unwrap());
```

(And then forgot about this refactoring.)

**And the code stopped to worked correctly**! I completely didn't understand what was going on. I debugged the code using gdb. And then after very time-consuming debug I suddenly understood that this refactoring changed meaning of the code.

The new code first reads `uncompressed` (in `uncompressed.split_at...`) and then mutates it (in `read_usize(&mut uncompressed)`). This is not what I want!

So, this is very big footgun. I propose to warn if same variable is read and is mutated in same expression. Of course, mutating same variable twice in one expression should warn, too.

(Ideally, rustc should warn on this code, too. Or even error. But this would be dream.)

### Example

```rust
// This is simplified example from my production code
#![feature(slice_take)]
fn main() {
    let uncompressed = vec![0, 0, 0];
    let mut uncompressed: &[u8] = &uncompressed;
    let (archive_info, uncompressed2) = uncompressed.split_at(usize::from(*<[u8]>::take_first(&mut uncompressed).unwrap()));
}
```

Could be written as:

```rust
#![feature(slice_take)]
fn main() {
    let uncompressed = vec![0, 0, 0];
    let mut uncompressed: &[u8] = &uncompressed;
    let tmp = uncompressed;
    let (archive_info, uncompressed2) = tmp.split_at(usize::from(*<[u8]>::take_first(&mut uncompressed).unwrap()));
}
```

or:

```rust
#![feature(slice_take)]
fn main() {
    let uncompressed = vec![0, 0, 0];
    let mut uncompressed: &[u8] = &uncompressed;
    let split_here = usize::from(*<[u8]>::take_first(&mut uncompressed).unwrap());
    let (archive_info, uncompressed2) = uncompressed.split_at(split_here);
}
```

depending on intended meaning

**Comment by blyxyas** [comment]
I think this is too important for Clippy to handle, should probably be moved to `rustc`

**Comment by Alexendoo** [comment]
For `rustc` it's probably too liable for false positives

e.g. if we cut the example down and replace the method call with function calls, it's clearer that `x` is evaluated before `write(&mut x)`

```rust
struct S;

fn read(_: &S, _: usize) {}
fn write(_: &mut &S) -> usize { 1 }

fn main() {
    let s = S;
    let mut x = &s;
    
    read(x, write(&mut x));
    //   ^ x is copied here, so there's no outstanding borrow to prevent the &mut x.
    //     That can be surprising for `&self` receiving methods
}
```

Relying on the evaluation order of a method call receiver being followed by the args could be worth a lint, it depends how much code makes use of that intentionally and how reliably we can detect it
